### PR TITLE
[streaming] New Source and state checkpointing interfaces 

### PR DESCRIPTION
**Pull request by StephanEwen**
These interfaces allow streaming sources/operators to interact with the state checkpointing in a more precise manner.

The interfaces are so far intended mainly for internal state handling (sources, window operators).

An open question what state interface we want to expose to the user code such that, on scale-in and scale-out, the state repartitions consistently with the data partitions.
  - For groupByKey operations, we could scope the state by key.
  - For all other operations, state may be repartitioned in a user-defined manner. This would give no guarantees about alignment with partitions. Think of a pipeline `partionBy -> map()`. The state accumulated in the `map()` function is partitioned by a key, but the state entries may not be able to be linked to a certain key or key-hash value.

### CODE REVIEW

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/checkpoint/Checkpointed.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	 *                            
+	 * @return A snapshot of the operator state.
+	 * 
+	 * @throws Exception Thrown if the creation of the state object failed. This causes the
+	 *                   checkpoint to fail. The system may decide to fail the operation (and trigger
+	 *                   recovery), or to discard this checkpoint attempt and to continue running
+	 *                   and to try again with the next checkpoint attempt.
+	 */
+	OperatorState<?> snapshotState(long checkpointId, long checkpointTimestamp) throws Exception;
+}
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/checkpoint/Checkpointed.java]: If the `checkpointTimestamp` is also derived at the JobManager why do we need both here? Maybe I am missing something. :)

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/StreamSource.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	 * 
+	 * @return True, if the end of the stream has been reached, false if more data is available.
+	 * 
+	 * @throws InterruptedException The calling thread may be interrupted to pull the function out of this
+	 *                              method during checkpoints.
+	 * @throws Exception Any other exception that is thrown causes the source to fail and results in failure of
+	 *                   the streaming program, or triggers recovery, depending on the program setup.
+	 */
+	boolean reachedEnd() throws Exception;
+
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/StreamSource.java]: Why did you choose to explicitly differ from the standard `hasNext()` name?

**Comment by mbalassi** [comment]
The changes seem reasonable, added comments inline.

As for the state elasticity my gut feeling is that we should be OK if we get the grouped case right. Besides that we do not provide partitioning guarantees for the records anyway (`partitionBy` is protected and only used when `groupBy` is called in `DataStream`).

**Comment by aljoscha** [comment]
I'm afraid we need to be a bit more strict here. We need to partition the state by key, to make it more manageable. Therefore, we should only allow the user to use state if he provides a key, for example using the (non-existant) keyBy(). Then we can partition the data and the operators know for which part of the state they are responsible.

**Comment by aljoscha** [comment]
Also, the key would be a property of a DataStream, and other operations could also use this.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/StreamSource.java (hunk 2 above)
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/StreamSource.java]: This is in line with the batch InputFormats, that is the only reason.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/checkpoint/Checkpointed.java (hunk 1 above)
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/checkpoint/Checkpointed.java]: The timestamp is the time when the checkpoint was started, assigned by the JobManager, part of the messages that the tasks receive.

All barriers emitted into the stream should to have the same timestamp, if we do consistent "source-assigned event time" - it makes triggering of time-tiggered windows much easier in operators.

Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/checkpoint/Checkpointed.java]: The checkpoint ID and the timestamp are also different. The ID is random unique and the timstamp is time

**Comment by StephanEwen** [comment]
@aljoscha I think this is orthogonal to the changes here. What you suggest is a more elaborate version of the `StateHandle`. For that, I totally agree.

**Comment by senorcarbone** [comment]
I like the interfaces, they are quite precise. To sum up (and see if I get this right), with `Checkpointed` we simply guarantee that an operator will not get invoked while state is persisted unless it implements `CheckpointedAsynchronously`. In that case the operator should instead send a copy of its state.

Regarding reconfiguration with arbitrary partitioning, a scheme that comes into my mind right now is to let functions to lazily request state on demand (without priorly injecting it) when they need to access it since we can't possible preallocate it. The first time they will try to access specific keys for example, only the associated state with these keys will be retrieved from the persistent recovery checkpoint. This way we let the reconfigured partitioners to decide on how state will be eventually allocated to instances.

**Comment by StephanEwen** [comment]
Concerning synchronous/asynchronous checkpointing, you are right. For asynchronous checkpointing, you may also use some copy-on-write techniques, rather than doing an eager copy.

For reconfiguring the state, we need still a way to have more fine grained handles to the state, to be able to split it in a way that is consistent with re-configuring the partitioner functions.

**Comment by gyfora** [comment]
Do you think these interfaces will work if at any point we want to do incremental snapshots?

**Comment by StephanEwen** [comment]
That is a good question. If we assume for now that tasking an incremental versus a full snapshot is a decision that each task takes individually, then I think it would work, as the checkpointing code does not see any difference.

The difference would be encoded in the state handle, which would have multiple fields, like `latestFullState` and `List<StateDelta>`. The operator can then reconstruct its full state form that.

**Comment by StephanEwen** [comment]
Sorry, my bad, I confused the layers.

These interfaces do not sufficiently describe things for incremental checkpointing. We may have to introduce additional interfaces for functions or operators that support incremental snapshotting.

**Comment by mbalassi** [comment]
I think these commits are in. @StephanEwen could you please close the PR manually?

**Comment by StephanEwen** [comment]
Manually merged!

**Pull request closed without merging**